src/comm.ts

```diff
@@ -0,0 +1,379 @@
+// Copyright (c) Jupyter Development Team.
+// Distributed under the terms of the Modified BSD License.
+"use strict";
+
+import { DisposableDelegate } from 'phosphor-disposable';
+
+import { 
+  IKernel, IKernelFuture, IKernelMessage, IKernelMessageOptions 
+} from './ikernel';
+
+import { createKernelMessage } from './kernel';
+
+import * as utils from './utils';
+
+
+/**
+ * A client side Comm interface.
+ */
+export
+interface IComm {
+  /**
+   * The uuid for the comm channel.
+   *
+   * Read-only
+   */
+  commId: string;
+
+  /** 
+   * The target name for the comm channel.
+   *
+   * Read-only
+   */
+  targetName: string;
+
+  /**
+   * The onClose handler.
+   */
+  onClose: (data?: any) => void;
+
+  /**
+   * The onMsg handler.
+   */
+  onMsg: (data: any) => void;
+
+  /**
+   * Open a comm with optional data.
+   */
+  open(data?: any, metadata?: any): IKernelFuture;
+
+  /**
+   * Send a comm message to the kernel.
+   */
+  send(data: any, metadata?: any, buffers?:(ArrayBuffer | ArrayBufferView)[]): IKernelFuture;
+
+  /**
+   * Close the comm.
+   */
+  close(data?: any, metadata?: any): IKernelFuture;
+}
+
+
+/**
+ * Contents of `comm_info` message.
+ */
+export
+interface ICommInfo {
+  /**
+   * Mapping of comm ids to target names.
+   */
+  comms: Map<string, string>;
+}
+
+
+/*
+ * CommManager for a Kernel.
+ *
+ * http://ipython.org/ipython-doc/dev/development/messaging.html#custom-messages
+ */
+export
+class CommManager {
+
+  /**
+   * Create a CommManager instance.
+   */
+  constructor(kernel: IKernel) {
+    this._kernel = kernel;
+    this._kernel.iopubReceived.connect(this._handleKernelMsg, this);
+  }
+
+  /**
+   * Connect to a comm, or create a new one.
+   *
+   * If a client-side comm already exists, it is returned.
+   */
+  connect(targetName: string, commId?: string): Promise<IComm> {
+    if (commId === void 0) {
+      commId = utils.uuid();
+    }
+    var promise = this._comms.get(commId);
+    if (promise) {
+      return promise;
+    }
+    var comm = new Comm(targetName, commId, this._kernel, () => {
+      this._unregisterComm(comm.commId);
+    });
+    promise = Promise.resolve(comm);
+    this._comms.set(commId, promise);
+    return promise;
+  }
+
+  /**
+   * Register the handler for a "comm_open" message on a given targetName.
+   */
+  registerTarget(targetName: string, cb: (comm: IComm, data: any) => void): void {
+    this._targets.set(targetName, cb);
+  }
+
+  /**
+   * Send a 'comm_info_request', and return the contents of the
+   * 'comm_info_reply'.
+   */
+  commInfo(targetName?: string): Promise<ICommInfo> {
+    var contents = {};
+    if (targetName !== void 0) {
+      contents = { target_name: targetName };
+    }
+    var future = sendCommMessage(this._kernel, 'comm_info_request', contents);
+    return new Promise((resolve, reject) => {
```

Comment:
It seems like we've lost some typing here. This *should* be a compile error, because `ICommInfo.comms` is typed as a `Map`, but it's really getting an object literal. So, I think `ICommInfo` definition should be updated, but we should see where we need to add an explicit type here to get this to fail as-is.